internal/pkg/service/common/etcdop/op/atomic_core.go

```diff
@@ -2,112 +2,217 @@ package op
+func (v *AtomicOpCore) RequireLock(lock Mutex) *AtomicOpCore {
+	v.locks = append(v.locks, lock)
 	return v
 }
 
-func (v *AtomicOpCore) Write(factories ...func(ctx context.Context) Op) *AtomicOpCore {
-	for _, fn := range factories {
-		v.WriteOrErr(func(ctx context.Context) (Op, error) {
-			return fn(ctx), nil
-		})
-	}
-	return v
+func (v *AtomicOpCore) newEmpty() *AtomicOpCore {
+	out := newAtomicCore(v.client)
+	out.checkPrefixKey = v.checkPrefixKey
+	return out
+}
+
+func (v *AtomicOpCore) setProcessorFactory(fn atomicOpCoreProcessorFactory) {
+	v.processorFactory = fn
 }
 
-func (v *AtomicOpCore) BeforeWrite(fns ...func(ctx context.Context)) *AtomicOpCore {
-	for _, fn := range fns {
-		v.WriteOrErr(func(ctx context.Context) (Op, error) {
-			fn(ctx)
-			return nil, nil
-		})
+func (v *AtomicOpCore) locksIfs() (cmps []etcd.Cmp, err error) {
+	for _, lock := range v.locks {
+		if key := lock.Key(); key == "" || key == "\x00" {
+			return nil, NotLockedError{}
+		}
+		cmps = append(cmps, lock.IsOwner())
 	}
-	return v
+	return cmps, nil
 }
 
-func (v *AtomicOpCore) BeforeWriteOrErr(fns ...func(ctx context.Context) error) *AtomicOpCore {
-	for _, fn := range fns {
-		v.WriteOrErr(func(ctx context.Context) (Op, error) {
-			return nil, fn(ctx)
-		})
+func (v *AtomicOpCore) readTxn(ctx context.Context, tracker *TrackerKV) (*TxnOp[NoResult], error) {
+	// Create READ transaction
+	readTxn := Txn(tracker)
+	for _, opFactory := range v.readPhase {
+		op := opFactory(ctx)
+		if errOp, ok := op.(*errorOp); ok {
+			return nil, errOp.err
+		} else if op != nil {
+			readTxn.Merge(op)
+		}
 	}
-	return v
+
+	// Stop the READ phase, if a lock is not locked
+	if lockIfs, err := v.locksIfs(); err == nil && len(lockIfs) > 0 {
+		readTxn.Merge(Txn(nil).
+			If(lockIfs...).
+			OnFailed(func(r *TxnResult[NoResult]) {
+				r.AddErr(errors.PrefixError(LockedError{}, "read phase"))
+			}),
+		)
+	} else if err != nil {
+		return nil, errors.PrefixError(err, "read phase")
+	}
+
+	return readTxn, nil
 }
 
-func (v *AtomicOpCore) WriteOp(ops ...Op) *AtomicOpCore {
-	for _, op := range ops {
-		v.Write(func(context.Context) Op {
-			return op
-		})
+func (v *AtomicOpCore) writeTxn(ctx context.Context) (*TxnOp[NoResult], error) {
+	// Create WRITE transaction
+	writeTxn := Txn(v.client)
+	for _, opFactory := range v.writePhase {
+		op := opFactory(ctx)
+		if errOp, ok := op.(*errorOp); ok {
+			return nil, errOp.err
+		} else if op != nil {
+			writeTxn.Merge(op)
+		}
 	}
-	return v
+
+	// Stop the WRITE phase, if a lock is not locked
+	if lockIfs, err := v.locksIfs(); err == nil && len(lockIfs) > 0 {
+		writeTxn.Merge(Txn(nil).
+			If(lockIfs...).
+			OnFailed(func(r *TxnResult[NoResult]) {
+				r.AddErr(errors.PrefixError(LockedError{}, "write phase"))
+			}),
+		)
+	} else if err != nil {
+		return nil, errors.PrefixError(err, "write phase")
+	}
+
+	// Attach processor callback, if any
+	var processor func(ctx context.Context, r *TxnResult[NoResult])
```

Comment:
Could we say that  `func(ctx context.Context, r *TxnResult[NoResult]` is `processorCallback`? I see this multiple times but actually cannot obtain the information what kind of function it is.

Reply:
I don't know if it makes sense, this is only one type of processor, there are more in the pkg.

In general, it is:
`func(ctx context.Context, r <RESULT>)`

So generic type can be `type processorFn[R any] func(ctx context.Context, r R)`

---------------

I see no benefit:

`var processor processorFn[TxnResult[NoResult]]`

vs 

`var processor func(ctx context.Context, r *TxnResult[NoResult])`

What do you think?

cc @jachym-tousek-keboola @hosekpeter